### PR TITLE
Generic deriving

### PR DESCRIPTION
**Pull request by zudov**
It should be quite ready now.

### CODE REVIEW

File: src/Data/Argonaut/Decode.purs [hunk 1]
```diff
@@ -37,6 +44,44 @@ import qualified Data.Map as Map
 class DecodeJson a where
   decodeJson :: Json -> Either String a
 
+-- | Decode `Json` representation of a value which has a `Generic` type.
+gDecodeJson :: forall a. (Generic a) => Json -> Either String a
+gDecodeJson = map (fromJust <<< fromSpine)
```
Comment (jdegoes) [inline src/Data/Argonaut/Decode.purs]: We should eliminate this `fromJust` and instead return a `Left` with some appropriate error message.

**Comment by jdegoes** [comment]
Looks great! Three minor nitpicks:

1. Can you squash commits?
2. The to / from list will incur a performance overhead, if you can directly fold over the map structure (at least for serializing to JSON), that will be faster.
3. It would be nice to bulk up the test data structure a bit to make the property tests more conclusive.

:+1: Thanks for your work on this!

File: src/Data/Argonaut/Decode.purs (hunk 1 above)
Comment (zudov) [inline src/Data/Argonaut/Decode.purs]: @jdegoes That's right. Fixed.

**Comment by zudov** [comment]
@jdegoes Nitpicks are welcome.

1. I'll squash everything after we review these last bits
2. See 60b2e29 and 57b2d72, not sure if that's the best what we can do without giving up safety but it already looks much better.
3. I can move data declarations a bit closer to the printing part, or to move printing part into a separate function and put it close to data declaration, is that what you meant?

**Comment by jdegoes** [comment]
(1) and (2) look fine. For (3), I meant, create a more complex data structure to exercise all the types exposed in Generic.

**Comment by zudov** [comment]
@jdegoes Makes sense. Fixed.

**Comment by jdegoes** [comment]
:+1: Awesome work! Can you squash commits one more time and I'll merge in?

**Comment by zudov** [comment]
Upps. Broke it with rebase, fixing now.

**Comment by zudov** [comment]
Ready to be merged now

**Comment by jdegoes** [comment]
:+1: Thanks again!

**Comment by zudov** [comment]
@jdegoes Thanks for the review

**Pull request merged**